Automattic/woocommerce-payments
# Implement payments via Express Checkout Element on the Cart and Checkout pages

Fixes https://github.com/Automattic/woocommerce-payments/issues/8870

## Changes proposed in this Pull Request









* Renders ECE elements on Shortcode Cart and Checkout pages when the `_wcpay_feature_stripe_ece` option is set to `1`.
* Renders ECE elements on the Product pages when the `_wcpay_feature_stripe_ece` option is set to `1`.


## Testing instructions





> [!IMPORTANT]
> This has been tested using [Google Pay](https://pay.google.com) since it is currently difficult to test Apple Pay due to Apple Pay's requirements for test cards.

> [!IMPORTANT]
> You must test these changes on a public facing URL using HTTPS. Use something like Jurassic Tube or ngrok to create a tunnel to your local environment.

### Enable the ECE elements

- (Docker) Open your local WCPay repository and run `./bin/cli.sh wp option update _wcpay_feature_stripe_ece 1`.
- (Valet) `cd` to your local merchant store **site** directory (**not** the WCPay repo) and run `wp option update _wcpay_feature_stripe_ece 1`.
    - **Note:** This requires WP CLI to be installed on your computer, you can install it with `brew install wp-cli`.
- (Everything) Open your database and add/update the `_wcpay_feature_stripe_ece` field in the `wp_options` table to be `1`.

### Make sure you have a Google Pay account set up with a real card

1. Create an account on [Google Pay](https://pay.google.com) if you don't already have one.
2. Make sure you have a real card added to your GPay account (it will not be charged while in test mode).
3. Make sure you're logged into GPay.
4. **Do all the tests in a Chromium based browser, such as Google Chrome, Arc, Vivaldi, etc.

### Testing procedure for cart and checkout pages

1. Add a product to your cart.
2. Go to the **Cart** page.
3. Make sure the Google Pay button is rendered.
4. Click the GPay button to initiate the GPay process.
5. Change the shipping address and make sure the total amount and shipping options are updated accordingly.
6. Change the shipping method and make sure the total amount is updated accordingly.
7. Click **Pay** and make sure the payment is processed successfully.

- Repeat these steps on both the shortcode and blocks based **Cart** and **Checkout** pages.
- Repeat once more with a product that does not require shipping.

### Test a failure message

> [!IMPORTANT]
> I don't know how to fake this because I don't know how/if we can use any GPay test cards, so there are some local code changes required here to fake an issue.

1. Set `phoneNumberRequired` to `false` here:

https://github.com/Automattic/woocommerce-payments/blob/1564ba9d8f0417fdfc76d3ac7ebdad35c1829ef1/client/express-checkout/index.js#L262

2. Follow steps (1) - (7) in the "Testing procedure for cart and checkout pages" above.
3. Make sure the payment fails.
4. Make sure there's an error notice displayed at the top of the page.

- Repeat these steps on both the shortcode and blocks based **Cart** and **Checkout** pages.

### Regression tests

1. Disable Apple Pay and Google Pay.
2. Test the payment critical flows with a credit card and make sure they work as intended.
3. Test WooPay and make sure it works as intended.

-------------------


**Post merge**



- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable. : _QA Testing Not Applicable_
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable. : _QA Testing Not Applicable_
